### ISSUE TITLE
Request: Allow nesting under selector without class/id name

### ISSUE THREAD

**Issue by paulozoom** [sass/sass#2030]
I was writing a mixin that would namespace a selector by prefixing it with a  breakpoint name. However, I also want to output a non-prefixed version, that works across all viewports. As I see it, this is a very common RWD situation.

Example:

``` scss
@include each-bp {
  &left { float: left; }
  &right { float: right; }
}
```

Output:

``` css
.left { float: left; }
.right { float: right; }

@media screen and (min-width: 500px) {
  .s-left { float: left; }
  .s-right { float: right; }
}
```

For this, I tried to write such a mixin:

``` scss
@mixin each-bp {
  // Unprefixed selector
  #{.} {
    @content;
  }
  // Prefixed selectors
  @each $bp, $size in $bps {
    @media screen and (min-width: $size) {
      .#{$bp}- {
        @content;
      }
    }
  }
}
```

However, the third line, `#{.} {`, isn’t valid Sass, because it expects a class name.

There isn’t, to my knowledge, an alternative for this that isn’t setting a global variable for the `#{$bp}-` prefix at each loop iteration and prepending it to each selector in the block passed to the mixin call, which is both inelegant — unnecessary pollutes the global namespace — and not very maintainable — changing the global variable’s name would imply an update to every selector in every call of this mixin.

Considering the lack of alternatives, and how common of a situation this is these days (see frameworks such as Bass, Intuit, Solid, etc.), I’d argue that for selectors with nested selectors, this requirement of a class name should be dropped, so this common use case becomes valid sass.

**Comment by cimmanon** [comment]
No, only bad designers write selectors like you're trying to write.  What you're trying to do is worse than "pollute[s] the global namespace", you're bloating your markup with unsemantic selectors with an added twist of having to modify your markup when your design changes.  You've created an unmaintanable mess in your CSS _and_ the markup.

The best case scenario for doing this in Sass is going to force you to generate a lot of selectors.  [Source for selector-nth](https://github.com/cimmanon/selector-nth).

``` scss
@import "selector-nth";

@function prefix-class($sel, $prefix) {
    @return #{str-slice($sel, 1, 1)}#{$prefix}-#{str-slice($sel, 2)};
}

@mixin each-bp($bps) {
    // Unprefixed selector
    & {
        @content;
    }

    // Prefixed selectors
    @each $bp, $size in $bps {
        @media screen and (min-width: $size) {
            @at-root #{selector-nth(&, -1, prefix-class, $bp)} {
                @content;
            }
        }
    }
}

$bps: (s, 500px),;
.foo, #bar {
    @include each-bp($bps) {
        float: left;
    }
}
```

Output:

``` css
.foo, #bar {
  float: left;
}
@media screen and (min-width: 500px) {
  .s-foo, #s-bar {
    float: left;
  }
}
```

In short, don't do this.

**Comment by paulozoom** [comment]
@cimmanon: I’d be happy to debate the pros and cons of developing with immutable utility classes, but that’s not what this request is about. Whether it makes sense for your/one’s projects or not, whether you/one likes it or not, it’s undeniable that **it is** a philosophy for CSS development, one already widely used in many responsive layout frameworks, and one that, I’d argue, is only growing in adoption. This is what the request is about: making Sass more useful for those who prefer to develop this way.

In short, let others do what makes sense for them.

**Comment by chriseppstein** [comment]
> In short, let others do what makes sense for them.

We get a lot of requests for `@include my-mixin($args) !important` which would then mark every property declaration in that mixin as important. We have rejected this feature not because it's hard but because it is bad to use `!important` like that. As such, we are frequently in the position of deciding whether language features in Sass encourage "Good CSS" or "Bad CSS" and we frequently have to use our own opinions on what constitutes good and bad stylesheet authoring practices. I'm not saying that this applies specifically here, just that as a general statement, I disagree with it :)

In this case, we want `&` to result in an error when there is no parent selector because we believe that most users expect there to be a parent selector in scope when they use it and so this error speeds up the debugging process of an authoring mistake.

My proposal to you is to change your implementation to use a prefix even for the global scope. This accomplishes all of your goals except for the exact naming you prefer. Alternatively, you can use the implementation suggestion that @cimmanon has made which seems pretty nice to me.

**Comment by paulozoom** [comment]
@chriseppstein:

> I'm not saying that this applies specifically here, just that as a general statement, I disagree with it :)

You're right, and as a general statement, I don’t agree with it myself 😄 

> […] we believe that most users expect there to be a parent selector in scope when they use it and so this error speeds up the debugging process of an authoring mistake

Totally understand that position of making life much easier on an use case that a lot more people will run into. Opened the issue with the hopes that both scenarios could be covered, but alas not. I’ll find a workaround.

As for @cimmanon’s suggestion, I appreciate it, but it’d imply calling the mixin for each selector group I want to namespace, resulting in one `@media` block per selector group, which isn’t desirable.
